src/rotator_library/token_calculator.py

```
@@ -0,0 +1,393 @@
+            None,
+            f"input_exceeds_context_by_{input_tokens - max_input_allowed}_tokens",
+        )
+
+    # Calculate available space for output
+    available_for_output = context_window - input_tokens - safety_buffer
+
+    if available_for_output < MIN_MAX_TOKENS:
+        # Input is too large - log warning but allow minimal response
+        logger.warning(
+            f"Input tokens ({input_tokens}) leave insufficient space for output "
+            f"(available: {available_for_output}, min: {MIN_MAX_TOKENS}). Model: {model}"
```

Ruff (0.15.4)

[warning] Logging statement uses f-string

(G004)